### ISSUE TITLE
Invalid SQL rendered for recursive query

### ISSUE THREAD

**Issue by denysvitali** [jOOQ/jOOQ#8966]
Hi everyone,
I'm trying to translate this (MariaDB / MySQL) query into in jOOQ:

```mysql
WITH RECURSIVE cte(role_id, inherits_from) as
(
	SELECT ri1.role_id, ri1.inherits_from
	FROM role_inheritance AS ri1
	WHERE ri1.role_id = 1
	UNION ALL
	SELECT ri2.role_id, ri2.inherits_from FROM role_inheritance AS ri2
	INNER JOIN cte
	ON ri2.role_id = cte.inherits_from
)

SELECT DISTINCT action, rule FROM permission WHERE owner_fk IN (
SELECT DISTINCT inherits_from FROM cte
INNER JOIN role as r1
ON r1.id = cte.role_id
INNER JOIN role as r2
ON r2.id = cte.inherits_from
UNION SELECT 1)
```

The tables involved are the followings:

```mysql
-- A role is a set of permissions
CREATE TABLE role (
    id INT AUTO_INCREMENT PRIMARY KEY,
    name VARCHAR(255) UNIQUE NOT NULL
);


CREATE TABLE role_inheritance (
  role_id INT,
  inherits_from INT,
  PRIMARY KEY (role_id, inherits_from),
  CONSTRAINT role_constraint
      FOREIGN KEY r1_fk(role_id)
          REFERENCES role(id)
          ON DELETE CASCADE,
  CONSTRAINT role_inherited_constraint
      FOREIGN KEY r2_fk(inherits_from)
          REFERENCES role(id)
          ON DELETE CASCADE
);

CREATE TABLE permission (
    id BIGINT PRIMARY KEY AUTO_INCREMENT,
    owner_fk INT NOT NULL,
    action VARCHAR(512) NOT NULL,
    rule ENUM('ALLOW', 'DENY') DEFAULT 'DENY',
    CONSTRAINT permission_constraint
    FOREIGN KEY (owner_fk)
    REFERENCES role(id)
);
```

And this is the jOOQ query:
```java
RoleInheritance ri1 = ROLE_INHERITANCE.as("ri1");
RoleInheritance ri2 = ROLE_INHERITANCE.as("ri2");

Role r1 = ROLE.as("r1");
Role r2 = ROLE.as("r2");

Select<Record2<String, PermissionRule>> r = dsl.withRecursive("cte", "role_id", "inherits_from")
		.as(select(ri1.ROLE_ID.as("role_id"), ri1.INHERITS_FROM.as("inherits_from"))
		.from(ri1)
		.where(ri1.ROLE_ID.eq(id))
		.unionAll(
			select(ri2.ROLE_ID, ri2.INHERITS_FROM)
			.from(ri2)
			.innerJoin(table(name("cte")))
			.on(ri2.ROLE_ID.eq(field(name("cte", "inherits_from"), Integer.class)))
		))
		.selectDistinct(PERMISSION.ACTION, PERMISSION.RULE)
		.from(PERMISSION)
		.where(PERMISSION.OWNER_FK.in(selectDistinct(field(name("cte", "inherits_from"), Integer.class))
				.from(table(name("cte"))
					.innerJoin(r1)
					.on(r1.ID.eq(field(name("cte","role_id"), Integer.class)))
					.innerJoin(r2)
					.on(r2.ID.eq(field(name("cte", "inherits_from"), Integer.class)))
				).union(select(inline(id)))
		));

System.out.println(r.getSQL());

return  r.fetch()
.stream()
.map(s -> s.into(PermissionRecord.class))
.collect(Collectors.toList());
```

The following returns:
```
with recursive `cte`(`role_id`, `inherits_from`) as ((select `ri1`.`role_id` as `role_id`, `ri1`.`inherits_from` as `inherits_from` from `project`.`role_inheritance` as `ri1` where `ri1`.`role_id` = ?) union all (select `ri2`.`role_id`, `ri2`.`inherits_from` from `project`.`role_inheritance` as `ri2` join `cte` on `ri2`.`role_id` = `cte`.`inherits_from`)) select distinct `project`.`permission`.`action`, `project`.`permission`.`rule` from `project`.`permission` where `project`.`permission`.`owner_fk` in ((select distinct `cte`.`inherits_from` from `cte` join `project`.`role` as `r1` on `r1`.`id` = `cte`.`role_id` join `project`.`role` as `r2` on `r2`.`id` = `cte`.`inherits_from`) union (select 1 from dual))
2019-07-18 13:48:29.666 DEBUG 8044 --- [nio-3000-exec-6] o.s.web.servlet.DispatcherServlet        : Failed to complete request: org.springframework.jdbc.BadSqlGrammarException: jOOQ; bad SQL grammar [with recursive `cte`(`role_id`, `inherits_from`) as ((select `ri1`.`role_id` as `role_id`, `ri1`.`inherits_from` as `inherits_from` from `project`.`role_inheritance` as `ri1` where `ri1`.`role_id` = ?) union all (select `ri2`.`role_id`, `ri2`.`inherits_from` from `project`.`role_inheritance` as `ri2` join `cte` on `ri2`.`role_id` = `cte`.`inherits_from`)) select distinct `project`.`permission`.`action`, `project`.`permission`.`rule` from `project`.`permission` where `project`.`permission`.`owner_fk` in ((select distinct `cte`.`inherits_from` from `cte` join `project`.`role` as `r1` on `r1`.`id` = `cte`.`role_id` join `project`.`role` as `r2` on `r2`.`id` = `cte`.`inherits_from`) union (select 1 from dual))]; nested exception is java.sql.SQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'union (select 1 from dual))' at line 1
```

From that jOOQ code, I expected the following SQL code:
```mysql
with recursive `cte`(`role_id`, `inherits_from`)
    as ((select `ri1`.`role_id`       as `role_id`,
    `ri1`.`inherits_from` as `inherits_from`
    from `project`.`role_inheritance` as `ri1`
    where `ri1`.`role_id` = 1)
    union all
    (select `ri2`.`role_id`, `ri2`.`inherits_from`
    from `project`.`role_inheritance` as `ri2`
     join `cte`
          on `ri2`.`role_id` = `cte`.`inherits_from`))
select distinct `project`.`permission`.`action`, `project`.`permission`.`rule`
from `project`.`permission`
where `project`.`permission`.`owner_fk` in (select distinct `inherits_from`
    from `cte`
             join `project`.`role` as `r1`
                  on `r1`.`id` = `cte`.`role_id`
             join `project`.`role` as `r2`
                  on `r2`.`id` = `cte`.`inherits_from`
   union
   select 1 from dual)
```

but got this instead
```mysql
with recursive `cte`(`role_id`, `inherits_from`)
    as ((select `ri1`.`role_id`       as `role_id`,
    `ri1`.`inherits_from` as `inherits_from`
    from `project`.`role_inheritance` as `ri1`
    where `ri1`.`role_id` = 1)
    union all
    (select `ri2`.`role_id`, `ri2`.`inherits_from`
    from `project`.`role_inheritance` as `ri2`
     join `cte`
          on `ri2`.`role_id` = `cte`.`inherits_from`))
select distinct `project`.`permission`.`action`, `project`.`permission`.`rule`
from `project`.`permission`
where `project`.`permission`.`owner_fk` in ((select distinct `inherits_from`
    from `cte`
             join `project`.`role` as `r1`
                  on `r1`.`id` = `cte`.`role_id`
             join `project`.`role` as `r2`
                  on `r2`.`id` = `cte`.`inherits_from`)
   union
   (select 1 from dual))
```

Basically, a parenthesis is added in the wrong place, thus the query becomes invalid.

I don't really know if this is a bug or it's just an user's fault (while translating from SQL to jOOQ). Can someone check please?

Thank you :)

**Comment by denysvitali** [comment]
P:S: Be careful with that query. If you edit it a bit you may go really deep with the recursion...
![screenshot](https://user-images.githubusercontent.com/4939519/61456039-6a328e80-a965-11e9-9b67-730270fdebe1.png)

**Comment by lukaseder** [comment]
Thank you very much for your report. We'll investigate this soon.

**Comment by lukaseder** [comment]
This seems to be yet another case where #3676 will fix the issues reported here. We're currently generating unnecessary parentheses around union all subqueries. These are necessary when such subqueries are nested, but they are not supported on MySQL 5.7. #3676 will fix this, and a backport to 3.12.4 was implemented.

**Comment by denysvitali** [comment]
Thank you! 🚀

**Comment by lukaseder** [comment]
Oh, I've just noticed: Your issue is about MariaDB, not MySQL 5.7. We cannot backport the fix for MariaDB to jOOQ 3.12, because in 3.12, we hadn't distinguished between MariaDB versions yet. However, if you build jOOQ 3.13.0-SNAPSHOT from Github, or if you have a commercial license and access the pre-built snapshot from here: https://www.jooq.org/download/versions, then the fix should be available already.
